### ISSUE TITLE
Error Breaking

### ISSUE THREAD

**Issue by joshrosario310** [joyfullservice/msaccess-vcs-integration#311]
Just want to start by saying how much I love this tool and how I just know it's going to change my life for the better, thank you very much!

I just fought with the add-in for several hours which was breaking for an error, I believe it was some form that didn't have any VBA. I noticed there were some On Error Resume Next lines nearby and finally realized I had inadvertently set my global VBA Editor to Break on All Errors. As soon as I set that back to Unhandled Errors, the add-in went back to exporting properly.

I don't know if it's possible to detect the User's Error Trapping setting before running the Export script but maybe it might save someone else some time down the road? I for sure will be looking there before any export operation going forward.

Unfortunately, it wasn't saving the log file after breaking but I think it was GetCodeModuleHash() in modHash that was throwing errors on exporting several specific forms.

**Comment by hecon5** [comment]
Check out [Options: General Tab](https://github.com/joyfullservice/msaccess-vcs-integration/wiki/Options#general-tab); thanks for the tip; been meaning to put together some more details on how the error thing works in the Wiki, but @joyfullservice has been forging ahead on the new merge tooling, so I've been waiting until that's settled down to do so.

**Comment by joyfullservice** [comment]
@joshrosario310 - Thanks for pointing this out! I did a little research this evening and found that the setting for this is stored in the registry in the following location:

![image](https://user-images.githubusercontent.com/7254992/160027230-f873700e-06ad-468e-bad5-a45708a739b4.png)

Those two values correspond to the setting saved by the current user for Error Trapping.

![image](https://user-images.githubusercontent.com/7254992/160027381-768f52f2-7b25-4976-8ee1-64d0715d2776.png)

I would propose that we add a check during the installation that verifies that the current user has VBA configured to _Break in Class Modules_. I can't really see a reason why any developer using this add-in would want it any other way. I know we have seen this before in other issues, so perhaps adding the check would help others avoid the hours of troubleshooting that Josh spent on it.

Although it would be trivial to change the registry value and transparently solve the problem for the user without them even knowing about it, I don't feel that this would be proper for the add-in to do this. Instead, I am proposing a message that warns them that they are not using the optimal setting for error handling in VBA, and that the add-in my not function as expected. This could then direct them to the wiki page as @hecon5 referenced, where they could find a more complete explanation of the setting. The user could then make an informed decision and change the setting themself in the VBA options dialog.

By way of explanation, this add-in uses error trapping directives in a number of places in the source code. There are certain things like determining whether a table has a table-level-macro that I have not found a way to identify without trapping the anticipate error. Other things like adding references to a database can generate errors, and it is better if we can trap those errors and provide meaningful information to the user rather than just breaking in the VBA source code and leaving users like Josh scratching their head trying to figure out what is wrong.

I will see if I can possibly put something together in the next few days to address this...

**Comment by joyfullservice** [comment]
For reference, I am linking to several other issues that related to this same _Error Trapping_ setting in the _VBA Options_. 

- #310 
- #269 
- #259 
- #220 
- #259 
- #66 
 
 In most of these cases, correcting this setting either resolved the problem, or helped them pinpoint exactly what was causing the issue. Adding the check during install seems like it would be a helpful enhancement to include in this add-in.

**Comment by Ik-12** [comment]
"Error Trapping" settings can be get/set in semi-documented way using Application.GetOption/SetOption functions, see e.g. https://www.fmsinc.com/free/NewTips/VBA/BasicErrorHandling.asp

My suggestion would be saving current "Error Trapping" state before export (or import), change it to appropriate value and then restore the original when export has finished. I use this method in my unit testing framework and it works flawlessly.

**Comment by joyfullservice** [comment]
> "Error Trapping" settings can be get/set in semi-documented way using Application.GetOption/SetOption functions, see e.g. https://www.fmsinc.com/free/NewTips/VBA/BasicErrorHandling.asp
> 
> My suggestion would be saving current "Error Trapping" state before export (or import), change it to appropriate value and then restore the original when export has finished. I use this method in my unit testing framework and it works flawlessly.

@Ik-12 - That is a fantastic idea! Thanks for pointing out the usage of `GetOption/SetOption` as well. I was not aware of this functionality and I like it much better than interacting directly with the registry values.

**Comment by hecon5** [comment]
ooh, that's really clean, I like it!

**Comment by joyfullservice** [comment]
I have implemented this concept in the `dev` branch to roll out in the version 4 release. 👍 

This concept of saving and restoring the session state also touched on the idea of cleaning up object references after the add-in completes the current operation. This is becoming increasingly important with the implementation of the ribbon interface, so this also lays some groundwork for this refactoring as well.

Whether an operation is launched from the add-in menu, the ribbon, or a call to the exposed COM API, we need to ensure that the environment is properly prepared for the add-in to run, then cleaned up afterwards so we don't have any dangling object references or temporarily adjusted user settings that could cause issues for our users.

We also need to account for the fact that in some scenarios the code to restore the user's settings may not run if the user ends the VBA code during a debug operation, or if the application crashes. For cases like this, the user's last saved settings should be restored the next time the add-in is run. (Commit coming soon for this.)

**Comment by hecon5** [comment]
@joyfullservice Since you've patched this, should this close, or do we need to handle more details on this?

**Comment by joyfullservice** [comment]
> @joyfullservice Since you've patched this, should this close, or do we need to handle more details on this?

Yes, I think we can close it now. I haven't done extensive testing with it, but it does seem to work for me. If we run into any further issues we can always reopen this issue, or create a new one. 👍
